### PR TITLE
fix for handling extClients when importing applications

### PR DESCRIPTION
**Pull request by gcornacchia**
when importing an application with an extclients credential, only the id had a value, not the clientid, so if you check the application in api manager web interface you see an empty text box. Also when you tried to re-import the same application an error occured, I saw that in the automated test there was no cases about extclients.

So I submitted this pull request to try to fix these things.

check it out

### CODE REVIEW

File: modules/apim-adapter/src/main/java/com/axway/apim/adapter/clientApps/APIMgrAppsAdapter.java [hunk 1]
```diff
@@ -374,19 +374,24 @@ private void saveCredentials(ClientApplication app, ClientApplication actualApp)
...
 			if(actualApp!=null && actualApp.getCredentials().contains(cred)) continue;
+			FilterProvider filter;
 			if(cred instanceof OAuth) {
 				endpoint = "oauth";
+				filter = new SimpleFilterProvider().setDefaultFilter(
+						SimpleBeanPropertyFilter.serializeAllExcept(new String[] {"credentialType", "clientId", "apiKey"}));
 			} else if (cred instanceof ExtClients) {
 				endpoint = "extclients";
+				filter = new SimpleFilterProvider().setDefaultFilter(
+						SimpleBeanPropertyFilter.serializeAllExcept(new String[] {"credentialType", "apiKey"}));
```
Comment (gcornacchia) [inline modules/apim-adapter/src/main/java/com/axway/apim/adapter/clientApps/APIMgrAppsAdapter.java:385]: I have to serialize the clientid in this case

File: modules/apim-adapter/src/main/java/com/axway/apim/api/model/apps/ClientApplication.java [hunk 2]
```diff
@@ -148,13 +157,14 @@ public boolean equals(Object other) {
 		if(other == null) return false;
 		if(other instanceof ClientApplication) {
 			ClientApplication otherApp = (ClientApplication)other;
+			Comparator c = Comparator.comparing(ClientAppCredential::getCredentialType).thenComparing(ClientAppCredential::getId);
```
Comment (gcornacchia) [inline modules/apim-adapter/src/main/java/com/axway/apim/api/model/apps/ClientApplication.java:160]: I have to sort the collections in order to determine if they are equals

File: modules/apps/src/test/resources/com/axway/apim/appimport/apps/basic/CompleteApplication.json [hunk 3]
```diff
@@ -22,7 +22,14 @@
     "secret" : "34f2b2d6-0334-4dcc-8442-e0e7009b8950",
     "corsOrigins" : [ "*" ],
     "apiKey" : "6cd55c27-675a-444a-9bc7-ae9a7869184d-${appNumber}"
-  } ],
+  } ,
+  {
+	    "credentialType" : "extclients",
```
Comment (gcornacchia) [inline modules/apps/src/test/resources/com/axway/apim/appimport/apps/basic/CompleteApplication.json:27]: adding the extclients credential in completeapplication.json for test

File: modules/apim-adapter/src/main/java/com/axway/apim/adapter/clientApps/APIMgrAppsAdapter.java (hunk 1 above)
Comment (cwiechmann) [reply modules/apim-adapter/src/main/java/com/axway/apim/adapter/clientApps/APIMgrAppsAdapter.java:385]: Have you tested this with oauth as well? I'm thinking, the clientId needs to be serialized as well for oauth, as it's using the same field. Do you had the change to test this as well?

File: modules/apim-adapter/src/main/java/com/axway/apim/api/model/apps/ClientApplication.java (hunk 2 above)
Comment (cwiechmann) [reply modules/apim-adapter/src/main/java/com/axway/apim/api/model/apps/ClientApplication.java:160]: Nice! Looks like I have to learn Java >8 stuff 😄

**Review by cwiechmann — COMMENTED** [review]
Very good happy to take that over. I have only one question concerning support for oauth.

**Comment by gcornacchia** [comment]
yes, I tested it, for OAuth credential what you see in web interface clientId is actually the id field in REST API. From what I see from the swagger the /app/{appid}/oauth API doesn't contain "clientId" field. It's only defined for extclient REST API

File: modules/apim-adapter/src/main/java/com/axway/apim/adapter/clientApps/APIMgrAppsAdapter.java (hunk 1 above)
Comment (gcornacchia) [reply modules/apim-adapter/src/main/java/com/axway/apim/adapter/clientApps/APIMgrAppsAdapter.java:422]: Yes, I tested it, in API Manager web interface what you see in clientId text field for oauth credential is the id field in REST API. Only the /extclients API defines the clientId field from what I saw in API Manager REST API swagger file.

Comment (cwiechmann) [reply modules/apim-adapter/src/main/java/com/axway/apim/adapter/clientApps/APIMgrAppsAdapter.java:422]: Perfect! Thanks for your confirmation. That I was no totally drunk as I have implemented this. 😃

**Review by cwiechmann — APPROVED** [review]

**Pull request merged**